### PR TITLE
CI/CD: Automatically publish Docker image

### PR DESCRIPTION
**Pull request by samsapti**
Now that #170 is fixed, we can add a CI/CD workflow to automatically push a Docker image to Dockerhub. Only thing you need to do is to add secrets for your Dockerhub credentials.

### CODE REVIEW

File: .github/workflows/docker-image.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
+name: Build and push Docker image
```
Comment (justinmayer) [inline .github/workflows/docker-image.yml:1]: Suggested re-wording: `Build and publish Docker image`

File: .github/workflows/docker-image.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+on:
+  push:
+    branches:
+      - main
+  workflow_dispatch:
```
Comment (justinmayer) [inline .github/workflows/docker-image.yml:7]: Nice work on including `workflow_dispatch`, so this Docker build+publish action can be [performed manually within GitHub's UI](https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/) 👍

File: .github/workflows/docker-image.yml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+jobs:
+  build-push:
```
Comment (justinmayer) [inline .github/workflows/docker-image.yml:10]: Suggested name for this job: `publish`

File: .github/workflows/docker-image.yml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    name: Build and push image
```
Comment (justinmayer) [inline .github/workflows/docker-image.yml:11]: Suggested re-wording: `Build and publish Docker image`

File: .github/workflows/docker-image.yml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    runs-on: ubuntu-latest
+    steps: 
+      - name: Login to DockerHub
```
Comment (justinmayer) [inline .github/workflows/docker-image.yml:14]: Suggested re-wording: `Log in to Docker Hub`

File: .github/workflows/docker-image.yml [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+        uses: docker/login-action@v2
+        with:
+          username: ${{ secrets.DOCKER_USERNAME }}
+          password: ${{ secrets.DOCKER_PASSWORD }}
+      
+      - name: Check out repository code
+        uses: actions/checkout@v2
+ 
+      - name: Build and push image
```
Comment (justinmayer) [inline .github/workflows/docker-image.yml:23]: Suggested re-wording: `Build and publish image`

**Review by justinmayer — APPROVED** [review]
Great work, Sam! I made a few *very* minor wording suggestions, just in case you find them useful. Excited to have this merged! ✨

**Comment by samsapti** [comment]
@justinmayer I like your suggestions :+1:

**Comment by samsapti** [comment]
I've also added a cron schedule to rebuild the image once a week. This is to keep the image (dependencies, etc.) up-to-date.

**Comment by lukevella** [comment]
Thanks for doing this! I’ll try and test this out asap but might have to wait till Monday.

> I've also added a cron schedule to rebuild the image once a week. This is to keep the image (dependencies, etc.) up to date.

I don’t think a cron schedule is necessary. The dependencies are locked with a lock file there shouldn’t be any difference in the dependencies if the commit is the same.

**Comment by samsapti** [comment]
> I don’t think a cron schedule is necessary. The dependencies are locked with a lock file there shouldn’t be any difference in the dependencies if the commit is the same.

@lukevella what I meant was for the Docker image itself to be kept up to date, meaning the base image (`node:alpine`) + Node.js, Yarn, etc. I can remove it if you want me to.

**Comment by lukevella** [comment]
> @lukevella what I meant was for the Docker image itself to be kept up to date, meaning the base image (`node:alpine`) + Node.js, Yarn, etc. I can remove it if you want me to.

Yes, I think it should be removed. Producing different results for the same commit would make debugging harder. Ideally, all dependency versions are pinned to avoid this from happening but that can be addressed separately.

**Comment by samsapti** [comment]
@lukevella done (force-pushed last commit but without cron schedule) :+1:

File: .github/workflows/docker-image.yml [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    runs-on: ubuntu-latest
+    steps: 
+      - name: Log in to Docker Hub
+        uses: docker/login-action@v2
+        with:
+          username: ${{ secrets.DOCKER_USERNAME }}
+          password: ${{ secrets.DOCKER_PASSWORD }}
+      
+      - name: Check out repository code
+        uses: actions/checkout@v2
```
Comment (valberg) [inline .github/workflows/docker-image.yml:20]: checkout is at v3 now, better use the newest version

**Review by valberg — CHANGES_REQUESTED** [review]

File: .github/workflows/docker-image.yml (hunk 7 above)
Comment (samsapti) [reply .github/workflows/docker-image.yml]: Fixed, thanks!

**Comment by lukevella** [comment]
Thanks, everyone for your help. Looks good to me!

**Pull request merged**